### PR TITLE
Implement categorical accessor

### PR DESCRIPTION
**Pull request by phofl**
This translates the logic from dask/dask, although I am not really happy with it.

I think it would be worth exploring to make stuff like ``categories``, ``as_known`` lazy as well. Unfortunately this has some drawbacks. Operations like ``pivot_table`` are using the categories to infer the structure of the result object. I don't have enough history here to judge whether this is something worth pursuing. Thoughts?

### CODE REVIEW

File: dask_expr/categorical.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        return self.set_categories(categories.values)
+
+    def as_unknown(self):
+        """Ensure the categories in this series are unknown"""
+        if not self.known:
+            return self._series
+
+        from dask_expr import new_collection
+
+        return new_collection(AsUnknown(self._series.expr))
```
Comment (mrocklin) [inline dask_expr/categorical.py:85]: Hrm, it's odd I think for Expr objects to return collections.  I think of Expr operations as being closed around Exprs.  Maybe some of this API belongs at the collection level and some of it belongs at the Expr level.  Maybe that's awkward though.  Thoughts?

File: dask_expr/categorical.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        ----------
+        kwargs
+            Keywords to pass on to the call to `compute`.
+        """
+        if self.known:
+            return self._series
+        from dask_expr.collection import new_collection
+
+        categories = (
+            new_collection(PropertyMap(self._series.expr, "cat", "categories"))
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply dask_expr/categorical.py:72]: It seems like anytime we have `new_collection(Operation(obj._expr, ...))` we should instead do `obj.operation(...)`

File: dask_expr/categorical.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                "supported.  Please use `column.cat.as_known()` or "
+                "`df.categorize()` beforehand to ensure known categories"
+            )
+            raise AttributeNotImplementedError(msg)
+        from dask_expr.collection import new_collection
+
+        return new_collection(PropertyMap(self._series.expr, "cat", "codes"))
+
+
+class AsUnknown(Blockwise):
```
Comment (mrocklin) [inline dask_expr/categorical.py:123]:
```suggestion
class AsUnknown(Elemwise):
```

File: dask_expr/acessor.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    -----
+    Subclasses should define ``_accessor_name``, ``_accessor_methods``, and
+    ``_accessor_properties``.
+    """
+
+    def __init__(self, series):
+        from dask_expr import Series
+
+        if not isinstance(series, Series):
+            raise ValueError("Accessor cannot be initialized")
```
[the remark this answers is not in the document]
Comment (phofl) [reply dask_expr/acessor.py:20]: Yeah that's why it's here. Yeah this shouldn't happen in user code

File: dask_expr/acessor.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+from dask.dataframe.accessor import _bind_method, _bind_property, maybe_wrap_pandas
```
[the remark this answers is not in the document]
Comment (phofl) [reply dask_expr/acessor.py:1]: thx

File: dask_expr/acessor.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    Base class for pandas Accessor objects cat, dt, and str.
+
+    Notes
+    -----
+    Subclasses should define ``_accessor_name``, ``_accessor_methods``, and
+    ``_accessor_properties``.
+    """
+
+    def __init__(self, series):
+        from dask_expr import Series
```
[the remark this answers is not in the document]
Comment (phofl) [reply dask_expr/acessor.py:17]: circular import unfortunately

File: dask_expr/categorical.py (hunk 1 above)
Comment (phofl) [reply dask_expr/categorical.py:85]: I am viewing the Accessors more as collections than as expressions. The accessor itself is registered on the Series collection

File: dask_expr/categorical.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (phofl) [reply dask_expr/categorical.py:72]: This is a bit of a chicken-egg issue here. 

```
self._series.cat.categories
```

accesses the categories of ``self._series._meta``, but these are unknown at this point. We have to trigger the actual computation to discover them

File: dask_expr/accessor.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    -----
+    Subclasses should define ``_accessor_name``, ``_accessor_methods``, and
+    ``_accessor_properties``.
+    """
+
+    def __init__(self, series):
+        from dask_expr import Series
+
+        if not isinstance(series, Series):
+            raise ValueError("Accessor cannot be initialized")
```
Comment (wence-) [inline dask_expr/accessor.py:20]: FWIW, to add type hints to this, you can do:

```python
from typing import  TYPE_CHECKING
if type_checking:
    from dask_expr import Series

...
class Accessor:
    def __init__(self, series: "Series"):
        ...
```

**Comment by phofl** [comment]
@mrocklin and I discussed this briefly last week and agreed to merge as is for now

**Pull request merged**